test/layout-test.ts

```diff
@@ -119,6 +251,64 @@ describe('layout', () => {
       assert.deepEqual(breakpoints, [0]);
     });
 
+    it('generates narrow frog prince layout from p. 81 of Digital Typography', () => {
+      const items = frogPrinceItems();
+      // width given on p. 78 of Digital Typography
+      // subtract 1em (18 machine units) from the first line
+      const lineLengths = [372, ...Array(items.length - 1).fill(390)];
+      const breakpoints = breakLines(items, lineLengths);
+      const lines = lineStrings(items, breakpoints);
+      assert.deepEqual(lines, [
+        'In olden times when wishing still helped one,',
+        'there lived a king whose daughters were all beau-',
+        'tiful; and the youngest was so beautiful that the',
+        'sun itself, which has seen so much, was aston-',
+        'ished whenever it shone in her face. Close by the',
+        'king’s castle lay a great dark forest, and under an',
+        'old limetree in the forest was a well, and when',
+        'the day was very warm, the king’s child went out',
+        'into the forest and sat down by the side of the',
+        'cool fountain; and when she was bored she took a',
+        'golden ball, and threw it up on high and caught',
+        'it; and this ball was her favorite plaything. -',
+      ]);
+      const adjRatios = adjustmentRatios(items, lineLengths, breakpoints).map((num) =>
+        Number(num.toFixed(3)),
+      );
+      assert.deepEqual(
+        adjRatios,
+        [0.857, 0.0, 0.28, 1.0, 0.067, -0.278, 0.536, -0.167, 0.7, -0.176, 0.357, 0.049],
```

Comment:
I notice the book image shows `0.000` as the final adjustment ratio rather than `0.049`. Do you know the reason for that? Does it matter at all?

Reply:
This is because you define `MAX_COST` to be 1000, but Knuth & Plass's calculations are based on a `MAX_COST` of 100,000. I chose to be consistent with the rest of the code in this repository, but if you want to be truly faithful to the book I can submit this PR with passing tests:

```typescript
diff --git a/test/layout-test.ts b/test/layout-test.ts
index 2a0f21d..1e0d274 100644
--- a/test/layout-test.ts
+++ b/test/layout-test.ts
@@ -232,9 +232,9 @@ function frogPrinceItems(): TextInputItem[] {
     }
   };
   const epilogue: TextInputItem[] = [
-    { type: 'penalty', width: 0, cost: 1000, flagged: false },
-    { type: 'glue', width: 0, stretch: 1000, shrink: 0, text: '' } as TextGlue,
-    { type: 'penalty', width: 0, cost: -1000, flagged: true },
+    { type: 'penalty', width: 0, cost: 100000, flagged: false },
+    { type: 'glue', width: 0, stretch: 100000, shrink: 0, text: '' } as TextGlue,
+    { type: 'penalty', width: 0, cost: -100000, flagged: true },
   ];
   return frogPrinceItemsImpl(frogPrinceText, prologue, betweenWords, epilogue);
 }
@@ -277,7 +277,7 @@ describe('layout', () => {
       );
       assert.deepEqual(
         adjRatios,
-        [0.857, 0.0, 0.28, 1.0, 0.067, -0.278, 0.536, -0.167, 0.7, -0.176, 0.357, 0.049],
+        [0.857, 0.0, 0.28, 1.0, 0.067, -0.278, 0.536, -0.167, 0.7, -0.176, 0.357, 0.0],
       );
     });
 
@@ -305,7 +305,7 @@ describe('layout', () => {
       );
       assert.deepEqual(
         adjRatios,
-        [0.774, 0.179, 0.629, 0.545, 0.0, 0.079, 0.282, 0.294, 0.575, 0.353],
+        [0.774, 0.179, 0.629, 0.545, 0.0, 0.079, 0.282, 0.294, 0.575, 0.004],
       );
     });
 
```

Reply:
> This is because you define MAX_COST to be 1000, but Knuth & Plass's calculations are based on a MAX_COST of 100,000.

Ah, thanks for the clarification. That's fine to leave as-is for the moment then.